pkg/controller/pxc/upgrade.go

```diff
@@ -115,7 +110,157 @@ func (r *ReconcilePerconaXtraDBCluster) updatePod(sfs api.StatefulApp, podSpec *
+	if !isPXC(sfs) {
+		return nil
+	}
+
+	if sfs.StatefulSet().Status.UpdatedReplicas >= sfs.StatefulSet().Status.Replicas {
+		return nil
+	}
+
+	if sfs.StatefulSet().Status.ReadyReplicas < sfs.StatefulSet().Status.Replicas {
+		return fmt.Errorf("failed to run 'SmartUpdate': not all replicas are ready")
+	}
+
+	list := corev1.PodList{}
+	err := r.client.List(context.TODO(),
+		&client.ListOptions{
+			Namespace:     sfs.StatefulSet().Namespace,
+			LabelSelector: labels.SelectorFromSet(sfs.Labels()),
+		},
+		&list,
+	)
+	if err != nil {
+		return fmt.Errorf("get pod list: %v", err)
+	}
+
+	log.Info("statefullSet was changed, run smart update")
+
+	primary, err := r.getPrimaryPod(cr)
+	if err != nil {
+		return fmt.Errorf("get primary pod: %v", err)
+	}
+
+	log.Info(fmt.Sprintf("primary pod is %s", primary))
+
+	waitLimit := 120
+	if cr.Spec.PXC.LivenessInitialDelaySeconds != nil {
+		waitLimit = int(*cr.Spec.PXC.LivenessInitialDelaySeconds)
+	}
+
+	sort.Slice(list.Items, func(i, j int) bool {
+		return list.Items[i].Name > list.Items[j].Name
+	})
+
+	var primaryPod *corev1.Pod
+	for _, pod := range list.Items {
+		pod := pod
+		if strings.HasPrefix(primary, fmt.Sprintf("%s.%s.%s", pod.Name, sfs.StatefulSet().Name, sfs.StatefulSet().Namespace)) {
+			primaryPod = &pod
+		} else {
+			log.Info(fmt.Sprintf("delete secondary pod %s", pod.Name))
+			err := r.client.Delete(context.TODO(), &pod)
+			if err != nil {
+				return fmt.Errorf("delete pod: %v", err)
+			}
+
+			err = r.waitPodRestart(&pod, waitLimit)
+			if err != nil {
+				return fmt.Errorf("wait pod: %v", err)
+			}
+		}
+	}
+
+	log.Info(fmt.Sprintf("delete primary pod %s", primaryPod.Name))
+	err = r.client.Delete(context.TODO(), primaryPod)
+	if err != nil {
+		return fmt.Errorf("delete primary pod: %v", err)
+	}
+
+	err = r.waitPodRestart(primaryPod, waitLimit)
+	if err != nil {
+		return fmt.Errorf("wait pod: %v", err)
+	}
+
+	return nil
+}
+
+func (r *ReconcilePerconaXtraDBCluster) getPrimaryPod(cr *api.PerconaXtraDBCluster) (string, error) {
+	user := "proxyadmin"
+	host := fmt.Sprintf("%s-proxysql-unready.%s", cr.ObjectMeta.Name, cr.Namespace)
+
+	var dbatabase queries.Database
+
+	for i := 0; ; i++ {
+		db, err := queries.New(r.client, cr.Namespace, cr.Spec.SecretsName, user, host, 6032)
+		if err != nil && i < int(cr.Spec.ProxySQL.Size) {
+			time.Sleep(time.Second)
+		} else if err != nil && i == int(cr.Spec.ProxySQL.Size) {
+			return "", err
+		} else {
+			dbatabase = db
+			break
+		}
+	}
+	defer dbatabase.Close()
+
+	return dbatabase.PrimaryHost()
+}
+
+func (r *ReconcilePerconaXtraDBCluster) waitPodRestart(pod *corev1.Pod, waitLimit int) error {
+	log.Info(fmt.Sprintf("wait pod %s start", pod.Name))
+	err := r.waitPodPhase(pod, corev1.PodPending, waitLimit)
+	if err != nil {
+		return err
+	}
+
+	err = r.waitPodPhase(pod, corev1.PodRunning, waitLimit)
+	if err != nil {
+		return err
+	}
+
+	log.Info(fmt.Sprintf("pod %s started", pod.Name))
```

Comment:
it is not enough to wait till `Running`; we should be sure that node in the `Synced` state after start

Reply:
https://github.com/percona/percona-xtradb-cluster-operator/pull/369/commits/84e9ccf1e9e0b87b09837b369b5170c22c4c4635